### ISSUE TITLE
UAC bypass with NIC (Native Image Cache) injection

### ISSUE THREAD

**Issue by AzAgarampur** [hfiref0x/UACME#78]
Hello,

I would like to contribute to this repository. (Basically add another method.) What are the steps I would need to take in order to add a new method?

**Comment by hfiref0x** [comment]
Hello,

can you describe this method first?

**Comment by AzAgarampur** [comment]
I have created a repository with the source and description. I have invited you as a collaborator. Please read `README.md` for a full description.

**Comment by hfiref0x** [comment]
I'm sorry but I do not collaborate with anyone. If you still want this viewed/added make it public - there is nothing zeroday in uac bypasses.

**Comment by AzAgarampur** [comment]
I respect that. This is the repo: https://github.com/AzAgarampur/byeintegrity-uac

**Comment by hfiref0x** [comment]
Can the target file be renamed so it can be replaced with own crafted? Or perhaps symlink? This is to minimize intrusion damage as I really don't want to patch any of Windows system files.

**Comment by AzAgarampur** [comment]
Yes. I have pushed new changes to the repository. The new change backs up the original system file.

**Comment by hfiref0x** [comment]
Okay I will look how to integrate this and do some tests, a week or so. Thanks.

**Comment by hfiref0x** [comment]
I could not reproduce this, clean 1809/2004, neither your code or my does not work as mmc never load this dll. Additionally desktop.ini in GAC folder results in 

SHCreateItemFromParsingName() (0) failed. HRESULT: 0x80070057

Do you have default Windows desktop.ini in the assembly folder?

**Comment by hfiref0x** [comment]
It seems it works only on 19041/19042 and maybe later, however it fails for example on insider builds. 
On clean 19042 EventVwr.ni.dll loaded by mmc.exe and there is no desktop.ini in GAC folder thus SHCreateItemFromParsingName succeeds.

**Comment by AzAgarampur** [comment]
OK here's what I got:
To answer your first question: No, there is no `desktop.ini` in GAC on build 19042.

I have only tested this on build 19042, and it seems to work. In order for this to work. To get this to work on older versions, .NET optimization task must have been executed at least once via auto system maintenance, otherwise `mmc.exe` does not load the native .NET images. It seems like 19042 has this done out of the box.

**Comment by hfiref0x** [comment]
It would be nice to figure out if this task can be triggered before exploitation otherwise it limits this method to only a newest Windows version with unstable conditions.

**Comment by AzAgarampur** [comment]
True, but remember it does not work right after a clean install. Most computers would have had the maintenance task ran at least once. I'm still trying to find a trigger that does not use UAC prompting.

**Comment by hfiref0x** [comment]
You mean \Microsoft\Windows\.NET Framework tasks? I've them all ran on test 18363 Windows and EventViewer.ni.dll is still not loaded by mmc.exe

**Comment by AzAgarampur** [comment]
Strange. I did the same and `EventViewer.ni.dll` seems to load after running the tasks... may have been a coincidence.

**Comment by AzAgarampur** [comment]
OK so I've updated the GitHub repo. On clean install of build 19042, it should continue to work fine. On earlier versions, the maintenance task needs to be run at least once. Windows 10 does not ship with `Desktop.ini` in the assembly folder by default, so this should work from 1st Windows 10 version to latest.

I also added a few changes that make it work from Win 7 if they do not have the `Desktop.ini` file. If they do have it, it will still fail.

I changed the target dll that is being attacked and instead of using the event viewer it now uses the firewall snap-in. I have ran the `\Microsoft\Windows.NET Framework` tasks and they do cause the new dll to be loaded into `mmc.exe`. The new dll is `Accessibility.ni.dll`.

**Comment by hfiref0x** [comment]
All my Win10 have this file up to 1809. The first Win10 without it seems 1903. That `mschedexe.exe` seems is autoelevated exe serving just as shell for calling task scheduler interface from `maintenanceui.dll`. Nice catch. Have you figured out what exactly blocks `SHCreateItemFromParsingName` from parsing files inside assembly folder?

**Comment by AzAgarampur** [comment]
Maybe the installation process creates the file... I simply opened `install.esd/wim` with 7-zip and looked inside the `Windows\assembly` directory and reported what I saw. As for `SHCreateItemFromParsingName`, I'm still trying to figure out why.

**Comment by hfiref0x** [comment]
That ini file is exactly what makes this dir look like special in explorer, just like recycle bin etc. So it maybe a Windows bug that it is missing this file in newest versions.

**Comment by AzAgarampur** [comment]
OK I did some experimenting with IDA and WinDbg and here's what I found: `SHCreateItemFromParsingName` -> `SHParseDisplayName` -> `windows.storage.dll!CFSFolder::ParseDisplayName` -> Loads `shfusion.dll.` -> `shfusion.dll!DllGetClassObject`.

I followed the returned class object and found it calls `shfusion.dll!CShellFolder::ParseDisplayName` -> **`return E_FAIL;`** The only thing that `shfusion.dll!CShellFolder::ParseDisplayName` does is immediately return `E_FAIL`. I am pretty sure this is causing our error.

This also explains why typing `C:\Windows\assembly\GAC` in file explorer results in an error when `Desktop.ini` exists under the assembly folder.

**Comment by hfiref0x** [comment]
I believe this is triggered after parsing CLSID indicated in that ini file -> `{1D2680C9-0E2A-469d-B787-065558BC7D43}`. It registry entry `HKEY_LOCAL_MACHINE\SOFTWARE\Classes\CLSID\{1D2680C9-0E2A-469d-B787-065558BC7D43}` has `Server` subkey with `ShFusion.dll` set as param.

It seems it is possible to redirect this behavior from machine wide setting to user setting by creating your own entry under 'HKEY_CURRENT_USER\SOFTWARE\Classes\CLSID\{1D2680C9-0E2A-469d-B787-065558BC7D43}` with removed `Server` parameter (just leave it empty). Then after restarting Explorer it will be able to browse `assembly` directory as normal directory. Unsure if this affects API immediatelly as I didn't yet tested this.

**Comment by AzAgarampur** [comment]
I tested this on my 19041. `SHCreateItemFromParsingName` succeeds immediately after creating the two registry keys and setting their values. However, `IFileOperation::MoveItem()` fails with the same error code. Restarting explorer does not help.

**Comment by AzAgarampur** [comment]
OK, in the remote `dllhost.exe` process, the function `windows.storage!CFIleOperation::_AddOperation` is being called. In that function, a function called `IsFolderItem(IShellItem*)` is being called. It returns `FALSE.` The function immediately returns `E_INVALIDARG`. This seems to be why `IFileOperation::MoveItem()` is failing.

**Comment by hfiref0x** [comment]
The problem is that without elevation you can't delete/modify this file. Maybe it need it own handler properly set in registry so shell will work as with usual directories.

**Comment by AzAgarampur** [comment]
There's a bug in `IFileOperation` that reports failure when the file copy actually has succeeded. With the registry change you posted, try to copy an empty `Desktop.ini` into `C:\Windows\assembly.` After confirming yes, it will attempt to copy and fail with an error. If you take a look at `C:\Windows\assembly\Desktop.ini`, you will see that it's actually empty. Now that `Desktop.ini` is empty, all `IFileOperation` methods work normally again.

I was thinking maybe we could do the same, while suppressing the conformation and elevation prompts. But I can't seem to get it working in my code.

However, I found a method that successfully works with `SHFileOperationW`, but it requires `SendInput`. I can't figure out how to get `SHFIleOperationW` to suppress the UAC prompt and I don't think you want that because it required GUI manipulation.

**Comment by AzAgarampur** [comment]
**Update:** Never mind, I was doing something wrong on my part. I got it to work. So now, the code will remove the `Desktop.ini` file and that allows `IFileOperation::MoveItem()` to work once again.

**Comment by AzAgarampur** [comment]
OK, the github repository is up to date with the latest code. Let's give this one a try.

**Comment by hfiref0x** [comment]
I've some idea how to get rid of IFileOperation completely so I want to do some tests. Btw do you know how exactly Windows manages these NI dlls in question to their file paths. I see they use sort of hash for directory naming. If we can figure out how to properly select correct path then search routines will be no longer needed.

**Comment by hfiref0x** [comment]
Okay, it succeeded here.

Instead of IFileOperation use simple Win32 API. Since this folder and file has full admin access set in DACL we can reset this file DACL to allow everyone any access for it. For this use ISecurityEditor autoelevated COM interface, previously abused by Simda malware. Basically this is autoelevated GetNamedSecurityInfo/SetNamedSecurityInfo.

Microsoft did some "fixes" in the past reducing this interface functionality to only SE_FILE_OBJECT's. Previously it took SE_* type as parameter and now it only works with files. However file is exactly what we need.

1. Build path to this NI dll (~this is seems last problem here)
2. ISecurityEditor->SetSecurity(FilePath, "D:(A;;GA;;;WD)") - this is sddl "all for everyone" set with DACL_SECURITY_INFORMATION
3. Open file normally, e.g. CreateFile
4. Patch it, or ovewrite completely (if overwrite then set new file size with SetEndOfFile)
5. Run target

**Comment by AzAgarampur** [comment]
Woah, that's awesome!! How did you figure that out? :) It will be much much easier using Win32 API for this.

As for building the path, all I know is that the "hash" folder name is a unique value to each version of the file. However, I've noticed that the folder with the latest creation time is always the one that is used. I am still trying to decode the meaning of the folder name.

**Comment by hfiref0x** [comment]
> How did you figure that out?

Sort of TRIZ.

There is a dirty way to find exact NI dll path. However it won't work with `wf.msc` snapin as it always require elevation. You can run mmc.exe with executionlevel asInvoker through appcompat environment variable (`__COMPAT_LAYER=RUNASINVOKER`). Once it will be loaded (for example `eventvwr.exe/eventvwr.msc`) it is possible to read mmc.exe memory (walk loader list/toolhelp whatever) and locate `EventViewer.Ni.dll` full path from it. Then terminate this process, modify file, run mmc normally. However I would prefer to generate filename using some algo instead.

**Comment by AzAgarampur** [comment]
Yup. I thought of the same thing. I can try to monitor the CLR and figure out the logic it uses to find the correct dll path.

**Comment by hfiref0x** [comment]
Just a note: in my test handle to this Accessibility.ni.dll is held by two VS processes `ServiceHub.Data.WarehouseHost.exe` and `ServiceHub.TestWindowStoreHost.exe`, so terminate them before calling CreateFile to avoid sharing violation.

**Comment by AzAgarampur** [comment]
OK so here's what I found out. The folder name is actually the debug GUID of the `Accessibility.ni.dll` file inside. However, that's not helpful to us.

To find which file to use, it first reads the non-NI file at `C:\Windows\Microsoft.NET\assembly\GAC_MSIL\Accessibility\v4.0_4.0.0.0__b03f5f7f11d50a3a\Accessibility.dll`. It will read the MVID (Module Version ID) of the file. Then, it recursively iterates through `C:\Windows\assembly\NativeImages_v4.0.30319_64\Accessibility` and reads each `.dll` file in each folder. The `dll` that has the matching MVID of `C:\Windows\Microsoft.NET\assembly\GAC_MSIL\Accessibility\v4.0_4.0.0.0__b03f5f7f11d50a3a\Accessibility.dll` is the file that is then loaded.

**Comment by hfiref0x** [comment]
I assume to figure out location of Accessibility.dll it is enough to enumerate registered assemblies via CreateAssemblyEnum/CreateAssemblyCache.

**Comment by AzAgarampur** [comment]
Does `CreateAssemblyEnum` give us the full path of the assembly? If it does, that's great. Otherwise, we will still have to manually enumerate through the `C:\Windows\Microsoft.NET\assembly` folder.

**Comment by hfiref0x** [comment]
Yes, IAssemblyCache->QueryAssemblyInfo will give full path to the registered assembly. Here is a test code https://gist.githubusercontent.com/hfiref0x/f9af768fc252f4d1ed91a67ec6997746/raw/a7de4eb89c488fbde8d5ba431a3266151951abab/fusion.c

**Comment by AzAgarampur** [comment]
Awesome 👍

Are we still going to start the automatic maintenance task? I don't think we need to because most systems would have ran the task at least once. The only case where they may not have ran it at least once is on a fresh Windows installation.

**Comment by hfiref0x** [comment]
Why not, I don't see any problem here. This method will be complicated in implementation even without this.

**Comment by AzAgarampur** [comment]
That is true, the only thing is running the task will run `ngen` which will take some time to finish. Also we have to monitor when the task finishes, because `mschedexe.exe` will exit immediately after launch.

**Comment by hfiref0x** [comment]
That is kind of impossible(?). Interface ignites a lot of additional executables and their life time depends on anything.
![Untitled](https://user-images.githubusercontent.com/10708977/93009733-93bd6900-f5ae-11ea-9d35-c661f6532853.png)

Perhaps some reasonable timeout is enough.

**Comment by AzAgarampur** [comment]
Sorry, I meant to say we cannot depend on `WaitForSingleObject` to wait on `mschedexe.exe` to exit and then assume that `ngen` task is finished. 

> Perhaps some reasonable timeout is enough.

That's what I was also leaning towards.
Perhaps we can use `NtQuerySystemInformation(SystemProcessInformation, ...)` in a loop with a 1.5 second delay to check if the `ngen.exe` task has finished?

**Comment by hfiref0x** [comment]
I see it runs ngen multiple times. Which one we should wait then?

**Comment by AzAgarampur** [comment]
![image](https://user-images.githubusercontent.com/42058942/93010020-751e8a00-f53c-11ea-822b-2d5ae929d1f2.png)

How about querying until these two processes do not exist anymore?

**Comment by hfiref0x** [comment]
Then comes another question, when we should check them? For example cleanmgr/defrag tasks are executed first, it could take a minutes for them to complete.

**Comment by AzAgarampur** [comment]
Well,

1) The task starts around 2 seconds after maintenance has started. We can start the query loop then.
2) You can delete any old versions of `C:\Windows\Microsoft.NET\Framework\v4.0.30319\ngen.log`, and when the `ngen` task has finished, the last line of the log file will contain `09/12/2020 21:26:03.699 [7920]: NGen task completed successfully`.

**Comment by hfiref0x** [comment]
Then we can remember this file last change, execute task and watch this file last change time with loop until it won't change (with exit on too many attempts, lets say > 5 min).

**Comment by AzAgarampur** [comment]
Sounds good. Btw `ngen` task finished early when the system is running on battery power and when battery level is low or critical.

**Comment by hfiref0x** [comment]
It seems this file is modified by few copies of ngen/ngentask with Read/Write share and WriteFile/CloseHandle updating it change time.

**Comment by AzAgarampur** [comment]
Will that be a problem? As long as the last write/change time has not changed for a set amount of time our wait mechanism should work. We can also use `NtQuerySystemInformation(SystemProcessInformation, ...)` as a last check to check for any `ngentask.exe` before we exit the wait mechanism.

**Comment by hfiref0x** [comment]
Windows 7 does not have `mschedexe.exe` and `fusion.dll` seems need to be loaded from different location `(Windows\Microsoft.NET\Framework64\v2.0.50727)`, unsure about that. On Win8/8.1 all the same as on 10.

**Comment by AzAgarampur** [comment]
Wait, I'm confused. Can you elaborate?

Btw Windows 7 uses the `Accessibility.ni.dll` out of the box on a fresh installation, no task is required.

**Comment by hfiref0x** [comment]
This dll exports CreateAssemblyCache which is used to determine full path to Accessibility.dll

**Comment by AzAgarampur** [comment]
OK, thanks for reminding me. I understand now.

You're right. Win. 7 has only `v2.0.50727` installed so we must use that if we're running on Windows 7.

**Comment by AzAgarampur** [comment]
Tiny problem: On my Windows 7, the call to `ShellExecuteW` is failing. It's returning Win32 code `8`, out of memory. It makes no sense as the system has enough resources available. Can you verify if this problem is happening to you as well?

**Comment by hfiref0x** [comment]
I have not experienced this issue. Last commit including working method, if there is no bugs found it will go to the master branch. I've tested this method on x64 Windows 7 SP1/8.1 full patch/10 (1809/2004) full patch and various Windows 10 Insider 21H1 Previews and it works well. I won't be implementing this for x32 as I see no sense it that.

**Comment by AzAgarampur** [comment]
Yeah, I reinstalled my Win 7 VM after leaving that comment and it started to work again... I'm not going to spend the time to try and figure out why.

OK, I'll look over the latest commit and let you know anything if I need to.

**Comment by AzAgarampur** [comment]
@hfiref0x This looks great. I believe it's ready to be merged with the master branch and included in the next release. (Also, why is your name is not in the `Authors` area of the readme? You helped a lot during the whole process.)

Thank you for the time you spent working with me and incorporating my method into UACME. I really appreciate it. 👍

**Comment by hfiref0x** [comment]
I usually put only original author name regardless of who and how implemented this. I will merge this later after some little polishing.

**Comment by AzAgarampur** [comment]
Awesome. Thank you!

**Comment by hfiref0x** [comment]
Done.
